qkernel/src/interrupt/aarch64/mod.rs

```diff
@@ -331,77 +336,49 @@ pub fn GetFaultAccessType(esr:u64, is_exe:bool) -> AccessType{
-            //
-            //TODO: on work
-            //
-            let error_code = PageFaultErrorCode::new(true, esr);
-            PageFaultHandler(ptregs , far, error_code);
-            return;
+
+pub fn HandleMemAbort(ptregs_addr: usize, esr: u64, far: u64, is_instr: bool, is_user: bool) {
+    debug!(
+        "get {} abort fault from {}, esr: {:x}",
+        match is_instr {
+            true => "instruction",
+            false => "data",
         },
-        _ => {
-            // TODO insert proper handler
-            panic!("DFSC/IFSC == 0x{:02x}, FAR == {:#x}, acces-type fault = {},\
-                   during address translation: {} ",  dfsc, far, access_type.String(),
-                   match EsrDefs::IsCM(esr) {
-                       true => "Yes",
-                       false => "No",
-                   });
+        match is_user {
+            true => "el0",
+            false => "el1",
         },
-    }
-}
-
-pub fn MemAbortKernel(ptregs_addr:usize, esr:u64, far:u64, is_instr:bool){
-    debug!("get {} abort fault from el1",
-           match is_instr {
-               true  => "instruction",
-               false => "data",
-           });
+        esr
+    );
+    use self::fault::PageFaultErrorCode as PFEC;
     let dfsc = esr & EsrDefs::ISS_DFSC_MASK;
     let access_type = GetFaultAccessType(esr, is_instr);
-    let dfsc_root = dfsc & PageFaultErrorCode::GEN_xxSC_MASK;
-    match dfsc_root {
-        PageFaultErrorCode::GEN_PERMISSION_FAULT |
-        PageFaultErrorCode::GEN_TRANSLATION_FAULT|
-        PageFaultErrorCode::GEN_ACCESS_FLAG_FAULT => {
-            info!("DFSC/IFSC == {:#X}, FAR == {:#X}, acces-type fault == {}, \
-                  during address translation == {}.", dfsc, far,
-                  access_type.String(), EsrDefs::IsCM(esr));
-            let ptregs_ptr = ptregs_addr as *mut PtRegs;
-            let ptregs = unsafe {
-                &mut *ptregs_ptr
-            };
-            let error_code = PageFaultErrorCode::new(false, esr);
-            PageFaultHandler(ptregs , far, error_code);
-            return;
-        },
-        _ => {
-            // TODO insert proper handler
-            panic!("DFSC/IFSC == 0x{:02x}, FAR == {:#x}, acces-type fault = {},\
-                   during address translation: {} ",  dfsc, far, access_type.String(),
-                   match EsrDefs::IsCM(esr) {
-                       true => "Yes",
-                       false => "No",
-                   });
-        },
+    let error_code = PFEC::new(is_user, is_instr, esr);
+    debug!("Page Fault, type: {:?}", error_code);
+    // early panic on faults that won't be handled
+    // TODO kill user instead of panic
+    // actually, this function should do no more than preparing parameters for
+    // the actual handler. if we decide to kill or panic, do it there, not here
+    if !error_code
+        .get_type()
+        .intersects(PFEC::PROTECTION_VIOLATION | PFEC::TRANSLATION | PFEC::ACCESS_FLAG)
```

Comment:
This slightly breaks the semantics, as it doesn't rule out a misconfigured code where more than one type flag is set. However this is logically impossible unless you intentionally do so.